platforms/ios/framework/src/TGMapView.mm

```diff
@@ -1,44 +1,318 @@
+    _rotationGestureRecognizer = [[UIRotationGestureRecognizer alloc] initWithTarget:self action:@selector(respondToRotationGesture:)];
+    _shoveGestureRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(respondToShoveGesture:)];
+    _shoveGestureRecognizer.minimumNumberOfTouches = 2;
+    _longPressGestureRecognizer = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(respondToLongPressGesture:)];
+
+    // Use the delegate method 'shouldRecognizeSimultaneouslyWithGestureRecognizer' for gestures that can be concurrent
+    _panGestureRecognizer.delegate = self;
+    _pinchGestureRecognizer.delegate = self;
+    _rotationGestureRecognizer.delegate = self;
+
+    [_glView addGestureRecognizer:_tapGestureRecognizer];
+    [_glView addGestureRecognizer:_doubleTapGestureRecognizer];
+    [_glView addGestureRecognizer:_panGestureRecognizer];
+    [_glView addGestureRecognizer:_pinchGestureRecognizer];
+    [_glView addGestureRecognizer:_rotationGestureRecognizer];
+    [_glView addGestureRecognizer:_shoveGestureRecognizer];
+    [_glView addGestureRecognizer:_longPressGestureRecognizer];
+}
+
+- (void)displayLinkUpdate:(CADisplayLink *)sender
+{
+    if (self.renderRequested || self.continuous) {
+        self.renderRequested = NO;
+        [self.glView display];
+    }
+}
+
+
+- (void)didReceiveMemoryWarning
+{
+    if (self.map) {
+        self.map->onMemoryWarning();
+    }
+}
+
+- (void)didMoveToWindow
+{
+    [self setupDisplayLink];
+    [super didMoveToWindow];
+}
+
+- (void)didMoveToSuperview
+{
+    [self setupDisplayLink];
+    [super didMoveToSuperview];
+}
+
+- (void)layoutSubviews
+{
+    [super layoutSubviews];
+
+    CGSize size = self.bounds.size;
+    self.map->resize(size.width * self.contentScaleFactor, size.height * self.contentScaleFactor);
+
+    [self requestRender];
+}
+
+- (void)requestRender
+{
+    if (!self.map) { return; }
+
+    self.renderRequested = YES;
+}
+
+- (void)captureScreenshot:(BOOL)waitForViewComplete
+{
+    self->_captureFrameWaitForViewComplete = waitForViewComplete;
+    self->_shouldCaptureFrame = YES;
+}
+
+- (void)glkView:(GLKView *)view drawInRect:(CGRect)rect
+{
+    if (self->_viewInBackground) {
+        return;
+    }
+
+    CFTimeInterval dt = _displayLink.targetTimestamp - _displayLink.timestamp;
+    _viewComplete = self.map->update(dt);
+
+    self.map->render();
+
+    if (_viewComplete && [self.mapViewDelegate respondsToSelector:@selector(mapViewDidCompleteLoading:)]) {
+        [self.mapViewDelegate mapViewDidCompleteLoading:self];
+    }
+
+    if (self.mapViewDelegate && [self.mapViewDelegate respondsToSelector:@selector(mapView:didCaptureScreenshot:)]) {
+        if (self->_shouldCaptureFrame && (!self->_captureFrameWaitForViewComplete || self->_viewComplete)) {
+            UIGraphicsBeginImageContext(self.frame.size);
+            [self drawViewHierarchyInRect:self.frame afterScreenUpdates:YES];
```

Comment:
Not required for this, but an FYI there's a `snapshot` property on GLKView that might be more performant. The downside is you will lose any UIKit modifications people may have added to the map view.

Reply:
Yeah I had the same thought! It's conceivable that you might want a screenshot of just the map canvas without decorations. We could just make that an option, like `withDecorations:(BOOL)`.